gateway/handle_http_auth.go

```diff
@@ -119,6 +123,57 @@ func (gw *Handle) sourceIDAuth(delegate http.HandlerFunc) http.HandlerFunc {
 	}
 }
 
+// authDestIDForSource middleware to authenticate destinationId in the X-Rudder-Destination-Id header.
+// If the destinationId is invalid, the request is rejected.
+// destinationID authentication should be performed only after source is authenticated and source is present in context
+// Following validations are performed
+//  1. Destination should be present for source config
+//  2. Destination should be enabled for the source
+func (gw *Handle) authDestIDForSource(delegate http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var errorMessage, reqType string
+		var arctx *gwtypes.AuthRequestContext
+		defer func() {
+			gw.handleHttpError(w, r, errorMessage)
+			gw.handleFailureStats(errorMessage, reqType, arctx)
+		}()
+		reqType, ok := gwCtx.GetRequestTypeFromCtx(r.Context())
+		if !ok {
```

Comment:
Now we need test scenarios for the `!ok` conditions :)

Reply:
Adding tests should be fine this way we will be able to prevent panics in the server code. @fracasula rightly pointed out that this will be a safer way to extract fields from context. 
https://github.com/rudderlabs/rudder-server/pull/4234#discussion_r1430111046